### PR TITLE
MySQL single master, multi-slave

### PR DESCRIPTION
**Pull request by jpoon**
This chart builds upon this [example](https://kubernetes.io/docs/tutorials/stateful-application/run-replicated-stateful-application/) to deploy a MySQL cluster with a single master that is replicated across multiple slaves.  

Tested with MySQL 5.7.13

### CODE REVIEW

**Comment by jpoon** [comment]
I signed it!

**Comment by jpoon** [comment]
I signed it!

**Comment by viglesiasce** [comment]
@jpoon could this be implemented as an extension to the existing mysql chart? Configurable through the values file?

File: incubator/mysqlha/templates/statefulset.yaml
Comment (viglesiasce) [inline incubator/mysqlha/templates/statefulset.yaml]: Since there is a lot going on in here can you split this out into its own file then use a Files.Get+ indent to add it in here?
There is an example in the spinnaker chart:
https://github.com/kubernetes/charts/blob/master/stable/spinnaker/templates/configmap/spinnaker-config.yaml#L11

File: incubator/mysqlha/templates/statefulset.yaml
Comment (viglesiasce) [inline incubator/mysqlha/templates/statefulset.yaml]: Make this configurable

File: incubator/mysqlha/templates/statefulset.yaml
Comment (viglesiasce) [inline incubator/mysqlha/templates/statefulset.yaml:161]: Same as above. Can you put this in its own file and then bring it in with Files.Get

File: incubator/mysqlha/values.yaml
Comment (viglesiasce) [inline incubator/mysqlha/values.yaml]: Do not set this by default in order to allow the default storage class of the cluster to be used.

File: incubator/mysqlha/templates/statefulset.yaml
Comment (viglesiasce) [inline incubator/mysqlha/templates/statefulset.yaml]: allow image repository to also be configurable

**Review by viglesiasce — CHANGES_REQUESTED** [review]
Was able to get the master to come up but then it kept crashing. The 2nd pod was stuck in a crash backoff loop:

```
mysql-ha-mysqlha-0                                   2/2       Running                 2          4m
mysql-ha-mysqlha-1                                   0/2       Init:CrashLoopBackOff   4          3m
```

**Comment by jpoon** [comment]
Thanks for the review @viglesiasce. I'm a lil overwhelmed at work at the moment and will address them as soon as I can.

**Comment by prydonius** [comment]
@jpoon gentle ping, are you able to pick this up again?

**Comment by jpoon** [comment]
@prydonius thanks for the reminder. been busy at work; hope to pick this back up and update the PR by mid-June.

**Comment by prydonius** [comment]
@jpoon no problem, since this is for the incubator, we could merge as-is so people can try it out/improve in the meantime. @viglesiasce we can visit the idea of merging this with stable/mysql when we move this to stable, would you be okay with merging as-is for now?

**Comment by prydonius** [comment]
@jpoon I think since the second pod is crashing we shouldn't merge this right now. Do you think you'll be able to pick this up now? If not I suggest we close this and you can always resubmit when you have the time

**Comment by jpoon** [comment]
Looks like this chart broke with k8s 1.6+, and the `init-containers` also went out of `beta` so there's some work to refactor the code. I'll close this PR and re-submit later.

**Comment by jpoon** [comment]
Thanks to @jonhoare for getting this work! Re-opening this PR.

**Comment by jonhoare** [comment]
CLA signed.

**Comment by jpoon** [comment]
Weird, the bot is still complaining. @jonhoare did you sign the CLA with the same email as the one configured in git `jon.bitbucket@livewire-uk.co.uk`?

**Comment by jonhoare** [comment]
You are right, I had signed up with my main email address jon@livewire-uk.co.uk, but I have now just signed into my linux identity account and modified the account email to be jon.bitbucket@livewire-uk.co.uk. So wonder if it has sorted itself out now?

**Comment by jonhoare** [comment]
I've now completely resigned up and signed a new CLA with this github account and jon.bitbucket@livewire-uk.co.uk, so I'm hoping the bot will be able to pick this up now? If this doesn't work I'll have to contact the helpdesk

**Comment by jonhoare** [comment]
It does't look like it has worked, so I have contacted the helpdesk.

**Comment by jonhoare** [comment]
Oh, ok, forget that it does seem to have now worked. 👍

File: incubator/mysqlha/templates/statefulset.yaml
Comment (jpoon) [reply incubator/mysqlha/templates/statefulset.yaml:161]: Spent an entire morning and couldn't get this to work with `Files.Get`.

**Comment by jpoon** [comment]
Couldn't get `Files.Get` to work, so I removed it. Deployed it on a 1.6.6 cluster and everything looks to be in order: pods are up, mysql is replicating.

**Comment by jpoon** [comment]
@viglesiasce @prydonius this PR is ready for your review.

(Thank you for your hard work in maintaining these charts)

**Comment by escardin** [comment]
@jpoon can you expose the MYSQL_DATABASE environment variable?

**Comment by escardin** [comment]
CLA Signed.

**Comment by thelinuxfoundation** [comment]
ping to bump for CLA check

**Comment by unguiculus** [comment]
Closing as stale. Feel free to reopen the PR if/when you decide to work on it again.

**Comment by jpoon** [comment]
@unguiculus the PR is ready for review.
@escardin did you sort out the CLA check? Seems like the bot still says you haven't signed it.

**Comment by escardin** [comment]
I think it might be failing the CLA check because of me. @jpoon if that is the case, I'll work with you to get my commits sorted out, or you can back them out, and when your bit is merged I can make a PR for my change.

**Comment by jpoon** [comment]
Weird, I don't have the rights to re-open this PR it looks like. 

@unguiculus is the CLA check the only thing that is blocking this from a review/merge? 
@escardin did you try emailing `helpdesk@rt.linuxfoundation.org`?

**Comment by escardin** [comment]
Yeah, I think I need to force push some commits with the right email on them, and then they need to get into your PR somehow.

**Comment by jpoon** [comment]
@escardin Force Push Fridays! I gave you push access to my fork (https://github.com/jpoon/charts/invitations)

File: incubator/mysqlha/templates/statefulset.yaml
Comment (escardin) [reply incubator/mysqlha/templates/statefulset.yaml:161]: I think this is now resolved

**Comment by escardin** [comment]
@jpoon I rebased, fixed my author and force pushed to your branch. Looks like the CLA is now good.
@viglesiasce I believe this is ready for re-review.

**Comment by viglesiasce** [comment]
Thanks for taking this on @escardin! Worked as expected for me.

**Review by viglesiasce — APPROVED** [review]

**Pull request merged**
